File: mc_raylib/src/camera_controller.rs
use raylib::prelude::*;

use std::f32::consts::PI;

use crate::consts as mn;

pub struct CameraController {
    pub camera: Camera3D,

    pub mouse_position: Vector2,
    pub mouse_sensitivity: f32,

    pub yaw: f32,   // left/right
    pub pitch: f32, // up/down
}
impl CameraController {
    pub fn new(camera: Camera3D) -> Self {
        Self {
            camera,

            mouse_position: Vector2::new(
                mn::WINDOW_WIDTH_START as f32 / 2.0,
                mn::WINDOW_HEIGHT_START as f32 / 2.0,
            ),
            mouse_sensitivity: mn::MOUSE_SENSITIVITY,

            yaw: 0.0,
            pitch: 0.0,
        }
    }
    pub fn move_by(&mut self, vec: Vector3) {
        self.camera.position += vec;
        self.camera.target += vec;
    }
    pub fn calc_forward(&self) -> Vector3 {
        let (yaw_sin, yaw_cos) = self.yaw.sin_cos();
        let (pitch_sin, pitch_cos) = self.pitch.sin_cos();

        Vector3::new(yaw_cos * pitch_cos, yaw_sin * pitch_cos, pitch_sin).normalized()
    }
    pub fn calc_right(&self) -> Vector3 {
        // TODO: maybe just use x/y of yaw instead?
        let (yaw_sin, yaw_cos) = (self.yaw - std::f32::consts::PI / 2.0).sin_cos();
        let (pitch_sin, pitch_cos) = self.pitch.sin_cos();

        Vector3::new(yaw_cos * pitch_cos, yaw_sin * pitch_cos, pitch_sin).normalized()
    }
    pub fn update(&mut self, rl: &mut RaylibHandle) {
        let mouse_position = rl.get_mouse_position();

        let mouse_delta = self.mouse_position - mouse_position;
        self.mouse_position = mouse_position;

        self.yaw += mouse_delta.x * self.mouse_sensitivity;
        self.pitch += mouse_delta.y * self.mouse_sensitivity;

        if self.pitch >= PI / 2.0 {
            self.pitch = PI / 2.0 - 0.001;
        }
        if self.pitch <= -PI / 2.0 {
            self.pitch = -PI / 2.0 + 0.001;
        }

        self.camera.target = self.camera.position + self.calc_forward();
    }
}

File: mc_raylib_cpp/cameraController.cpp
#include "raylib.h"

#include <cmath>

#include "include/consts.h"

#include "include/Vector3Util.h"
#include "include/Vector2Util.h"

#include "include/cameraController.h"


CameraController::CameraController() {
	this->camera = { 0 };
	this->camera.position =	(Vector3){ 0.0, 0.0, 0.0 };
	this->camera.target =	(Vector3){ 1.0, 0.0, 0.0 };
	this->camera.up = 		(Vector3){ 0.0, 0.0, 1.0 };
	this->camera.fovy =		60.0f;
	this->camera.projection = CAMERA_PERSPECTIVE;

	this->mousePosition = (Vector2){
		WINDOW_WIDTH_START / 2.0,
		WINDOW_HEIGHT_START / 2.0,
	};
	this->mouseSensitivity = MOUSE_SENSITIVITY;

	this->yaw = 0.0;
	this->pitch = 0.0;
}

void CameraController::moveBy(const Vector3 &vec) {
	this->camera.position += vec;
	this->camera.target += vec;
}

Vector3 CameraController::calcForward() const {
	float yawSin = sin(this->yaw);
	float yawCos = cos(this->yaw);

	float pitchSin = sin(this->pitch);
	float pitchCos = cos(this->pitch);

	return (Vector3) {
		yawCos * pitchCos,
		yawSin * pitchCos,
		pitchSin,
	};
}
Vector3 CameraController::calcRight() const {
	float rightYaw = this->yaw - PI / 2.0;
	float yawSin = sin(rightYaw);
	float yawCos = cos(rightYaw);

	float pitchSin = sin(this->pitch);
	float pitchCos = cos(this->pitch);

	return (Vector3) {
		yawCos * pitchCos,
		yawSin * pitchCos,
		pitchSin,
	};
}

void CameraController::update() {
	Vector2 newMousePosition = GetMousePosition();
	Vector2 mouseDelta = this->mousePosition - newMousePosition;
	this->mousePosition = newMousePosition;

	this->yaw += mouseDelta.x * this->mouseSensitivity;
	this->pitch += mouseDelta.y * this->mouseSensitivity;

	if (this->pitch >= PI / 2.0) {
		this->pitch = PI / 2.0 - 0.01;
	}
	else if (this->pitch <= -PI / 2.0) {
		this->pitch = -PI / 2.0 + 0.01;
	}
	
	this->camera.target = this->camera.position + this->calcForward();
}

tuple<int, int, int> CameraController::getChunkPos() const {
	return (tuple<int, int, int>) {
		(int)floor(this->camera.position.x / (float)CHUNK_SIZE),
		(int)floor(this->camera.position.y / (float)CHUNK_SIZE),
		(int)floor(this->camera.position.z / (float)CHUNK_SIZE),
	};
}